sql/src/main/java/org/apache/apex/malhar/sql/planner/RelNodeVisitor.java

```diff
@@ -0,0 +1,457 @@
+    }
+
+    if ((join.getJoinType() == JoinRelType.FULL) || (join.getJoinType() == JoinRelType.LEFT) ||
+        (join.getJoinType() == JoinRelType.RIGHT)) {
+      throw new UnsupportedOperationException("Outer joins are not supported");
+    }
+
+    final List<Integer> leftKeys = new ArrayList<>();
+    final List<Integer> rightKeys = new ArrayList<>();
+
+    RexNode remaining =
+        RelOptUtil.splitJoinCondition(join.getLeft(), join.getRight(), join.getCondition(), leftKeys, rightKeys);
+
+    if (leftKeys.size() != rightKeys.size()) {
+      throw new RuntimeException("Unexpected condition reached. Left and right condition count should be same");
+    }
+
+    if (leftKeys.size() == 0) {
+      throw new UnsupportedOperationException("Theta joins are not supported.");
+    }
+
+    RelInfo relInfo = addInnerJoinOperator(join, leftKeys, rightKeys);
+
+    if (!remaining.isAlwaysTrue()) {
+      relInfo = addJoinFilter(join, remaining, relInfo);
+    }
+
+    return relInfo;
+  }
+
+  private RelInfo addJoinFilter(Join join, RexNode remaining, RelInfo relInfo)
+  {
+    FilterTransformOperator operator = dag.addOperator(OperatorUtils.getUniqueOperatorName(join.getRelTypeName() +
+        "_Filter"), FilterTransformOperator.class);
+    ExpressionCompiler compiler = new ExpressionCompiler(new RexBuilder(join.getCluster().getTypeFactory()));
+    String expression = compiler.getExpression(remaining, join.getRowType(), join.getRowType());
+
+    Map<String, String> expMap = new HashMap<>();
+    for (Pair<RelDataTypeField, RelDataTypeField> pair : Pair.zip(join.getRowType().getFieldList(),
+        join.getRowType().getFieldList())) {
+      String leftName = OperatorUtils.getFieldName(pair.left);
+      String rightName = OperatorUtils.getFieldName(pair.right);
+      expMap.put(leftName, rightName);
+    }
+    operator.setExpressionMap(expMap);
+    operator.setCondition(expression);
+
+    String streamName = OperatorUtils.getUniqueStreamName(join.getRelTypeName() + "_Join", join.getRelTypeName() +
+        "_Filter");
+    Class schema = getSchemaForRelDataType(streamName, relInfo.getOutRelDataType());
+    dag.setOutputPortAttribute(relInfo.getOutPort(), Context.PortContext.TUPLE_CLASS, schema);
+    dag.setInputPortAttribute(operator.input, Context.PortContext.TUPLE_CLASS, schema);
+    dag.addStream(streamName, relInfo.getOutPort(), operator.input);
+
+    return new RelInfo("Join", relInfo.getInputPorts(), operator, operator.output, join.getRowType());
+  }
+
+  private RelInfo addInnerJoinOperator(Join join, List<Integer> leftKeys, List<Integer> rightKeys)
+  {
+    String leftKeyExpression = null;
+    String rightKeyExpression = null;
+    for (Integer leftKey : leftKeys) {
+      String name = OperatorUtils.getValidFieldName(join.getLeft().getRowType().getFieldList().get(leftKey));
+      leftKeyExpression = (leftKeyExpression == null) ? name : leftKeyExpression + " + " + name;
```

Comment:
@tushargosavi Order of the keys is preserved while comparison string is generated. i.e. leftKeyExpression will become "a + b"... Hence for the 2 cases that you mentioned.. The result of expression will be 1011 & 1110 which are different.
Please correct me if I've missing anything.